packages/lesswrong/server/vulcan-lib/apollo-ssr/renderPage.tsx

```diff
@@ -145,6 +159,39 @@ export const renderWithCache = async (req: Request, res: Response, user: DbUser|
   }
 };
 
+let inFlightRenderCount = 0;
+const requestQueue: RenderQueueSlot[] = [];
+
+/**
+ * We (maybe) have a problem where too many concurrently rendering requests cause our servers to fall over
+ * To solve this, we introduce a queue for incoming requests, such that we have a maximum number of requests being rendered at the same time
+ * See {@link maybeStartQueuedRequests} for the part that kicks off requests when appropriate
+ */
+function queueRenderRequest(params: RenderRequestParams): Promise<RenderResult> {
+  return new Promise((resolve) => {
+    requestQueue.push({
+      callback: async () => {
+        const result = await renderRequest(params);
+        inFlightRenderCount--;
```

Comment:
If `renderRequest()` throws an uncaught error this won't get decremented, which could mean that the server gets stuck unable to process any requests. I've tested and this actually does happen if you throw an error straight away in `renderRequest`, it also seems not unlikely to happen in real life if the server is being spammed with requests. Wrapping this in a `try ... finally` should fix this, better still would be keeping an array of inflight requests so you can always be sure you have the right count